examples/tv-casting-app/tv-casting-common/core/CastingPlayer.cpp

```diff
@@ -130,15 +173,66 @@ void CastingPlayer::VerifyOrEstablishConnection(ConnectCallback onCompleted, uns
-        mCommissioningWindowTimeoutSec = kCommissioningWindowTimeoutSec;
-        mTargetCastingPlayer           = nullptr;
-        mOnCompleted(err, nullptr);
-        mOnCompleted = nullptr;
+        resetState(err);
     }
 }
 
+void CastingPlayer::ContinueConnecting(ConnectionCallbacks connectionCallbacks,
+                                       unsigned long long int commissioningWindowTimeoutSec)
+{
+    ChipLogProgress(AppServer, "CastingPlayer::ContinueConnecting()");
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    SuccessOrExit(err = support::ChipDeviceEventHandler::SetUdcStatus(true));
+    VerifyOrExit(
+        connectionCallbacks.mOnConnectionComplete != nullptr,
+        ChipLogError(AppServer, "CastingPlayer::ContinueConnecting() ConnectionCallbacks.mOnConnectionComplete was not provided"));
+    mConnectionState               = CASTING_PLAYER_CONNECTING;
+    mOnCompleted                   = connectionCallbacks.mOnConnectionComplete;
+    mCommissioningWindowTimeoutSec = commissioningWindowTimeoutSec;
+    mTargetCastingPlayer           = this;
+
+    // Register the handler for Commissioner's CommissionerDeclaration messages. The CommissionerDeclaration messages provide
+    // information indicating the Commissioner's pre-commissioning state.
+    if (connectionCallbacks.mCommissionerDeclarationCallback != nullptr)
+    {
+        matter::casting::core::CommissionerDeclarationHandler::GetInstance()->SetCommissionerDeclarationCallback(
+            connectionCallbacks.mCommissionerDeclarationCallback);
+    }
+    else
+    {
+        ChipLogProgress(
+            AppServer,
+            "CastingPlayer::VerifyOrEstablishConnection() CommissionerDeclarationCallback not provided in ConnectionCallbacks");
+    }
+
+    ChipLogProgress(AppServer, "CastingPlayer::ContinueConnecting() calling OpenBasicCommissioningWindow()");
+    SuccessOrExit(err = chip::Server::GetInstance().GetCommissioningWindowManager().OpenBasicCommissioningWindow(
+                      chip::System::Clock::Seconds16(mCommissioningWindowTimeoutSec)));
+
+    mIdOptions.mCommissionerPasscodeReady = true;
+    ChipLogProgress(AppServer, "CastingPlayer::ContinueConnecting() calling SendUserDirectedCommissioningRequest()");
+#if CHIP_DEVICE_CONFIG_ENABLE_COMMISSIONER_DISCOVERY_CLIENT
+    SuccessOrExit(err = SendUserDirectedCommissioningRequest());
+#endif // CHIP_DEVICE_CONFIG_ENABLE_COMMISSIONER_DISCOVERY_CLIENT
+
+exit:
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(AppServer, "CastingPlayer::ContinueConnecting() failed with %" CHIP_ERROR_FORMAT, err.Format());
+        resetState(err);
+    }
+}
+
+void CastingPlayer::resetState(CHIP_ERROR err)
+{
+    support::ChipDeviceEventHandler::SetUdcStatus(false);
+    mConnectionState               = CASTING_PLAYER_NOT_CONNECTED;
+    mCommissioningWindowTimeoutSec = kCommissioningWindowTimeoutSec;
+    mTargetCastingPlayer           = nullptr;
+    mOnCompleted(err, nullptr);
```

Comment:
should we check that mOnCompleted is not null already? Or is that fine since we would segfault on invalid usage anyway?

Reply:
Adding check for null to avoid a potential segmentation fault since it's good practice. It is somewhat redundant since we check if the passed in connectionCallbacks.mOnConnectionComplete is not null above in:

    VerifyOrExit(
        connectionCallbacks.mOnConnectionComplete != nullptr,
        ChipLogError(AppServer, "CastingPlayer::ContinueConnecting() ConnectionCallbacks.mOnConnectionComplete was not provided"));